### ISSUE TITLE
Publish app in F-Droid

### ISSUE THREAD

**Issue by KinkyGeek** [fetlife/android#249]
There is a project called [F-Droid](https://f-droid.org) which is an open source app store for Android whereby only free and open source apps are allowed. Perhaps the FetLife app could be added to it?

**Comment by kistitan** [comment]
Hi @KinkyGeek,

Thank you for your suggestion. We are investigating options to add to app to one of the alternative App store.
F-Droid is one that we are also looking at.

At this moment however it does not seem to be a solution for us.
The problem with F-Droid is that not only build the app by themselves (what is a nice approach as it guarantees that the apk is indeeded based on the open source project) but they also sign it themselves.

For us it is a security issue. Even if the whole source code of the app is public we keep some keys, like the signing key private. The reason for it is that we want to give the users a way to make sure they are using the official FetLife app. This can prevent somebody starting the make a fraud FetLife app based on our code and spread it possibly stealing money or privacy information. 

If we let the F-Droid sign the app instead of us we would give the control of this very important security aspect of our hands.

As of now getjar and mikandi seem to be a better option for us.

We have an open issue for this discussion:
https://github.com/fetlife/android/issues/179

please feel free to add your comment, remarks there.

**Comment by Mikaela** [comment]
Wouldn't the signing key issue with F-Droid be resolved by running your own repository and instructing (or linking to instructions) how to add it to F-Droid client? https://f-droid.org/wiki/page/Setup_an_FDroid_App_Repo

**Comment by KinkyGeek** [comment]
I was going to comment and mention this feature of F-Droid, but forgot. Thanks @Mikaela The user experience (from within the F-Droid app) is pretty easy and nice as well.

Hopefully the kinky genius people at Fetlife will like this suggestion which will surely please the privacy conscious / nerd users like me :grin:

**Comment by kistitan** [comment]
Hmm...
First of thank you for helping us out with F-droid issue.
We really appreciate your help.

However unfortunately the signing issue can't be solve that way either.
Because to have the app signed by our key we either should give our private key to a third party place (we really don't want to do it) or we have to signe the compiled version by ourselves and after upload it to a place where F-droid can share. This last one seems to be against the distribution process of F-droid

**Comment by kistitan** [comment]
However as our app is open source there is nothing against the have a user built it by her or himself so if F-droid client supports this (I didn't take a deep look yet) we can indeed share some instructions how is the easiest way to do it.
Let me investigate this a bit later.

**Comment by KinkyGeek** [comment]
> Because to have the app signed by our key we either should give our private key to a third party place (we really don't want to do it)

@kistitan I do not think you understand, setting up an "F-Droid App Repo" allows projects to setup their own repository, on their own server, to publish their own builds precisely to avoid this signing key aspect. The documentation @Mikaela pointed to states in [Real World Setup](https://f-droid.org/wiki/page/Setup_an_FDroid_App_Repo#Real_World_Setup):

> "...repo signing keys should not ever be on a public server..."

and encourages the approach:

> "...generate the repo on a non-public machine like your laptop, keeping config.py and the keystore only on that machine"

**Comment by kistitan** [comment]
Hi @KinkyGeek,

Thank you for pointing out these part of the documentation. 
Indeed, as our main priority was to get into the official F-Droid repo (what is not possible with using our own key) I just quickly ran through the documentation above, without deeply understanding the suggestion. 
Sorry for that @Mikaela and @KinkyGeek.

So as far as I understand now the concept basically is to setup our own repo based on F-Droid server and publish our own app there, and that new repo you can add to your F-Droid client.
Sounds doable.

Give me some time to finish other tasks than come back to this again.

Thanks again.

**Comment by KinkyGeek** [comment]
@kistitan glad to hear it makes sense and might be doable. Thanks :)

**Comment by hex-m** [comment]
@kistitan an elegant solution to the security issue you mentioned would be to make fetlife [build reproducible](https://f-droid.org/docs/Reproducible_Builds/). The idea is that you create a recipe that can be used to create an identical apk from the source code.

> Publishing signed binaries from elsewhere (e.g. the upstream developer) is now possible after verifying that they match ones built using a recipe.

**Comment by Bubu** [comment]
F-Droid contributor here, a few thoughts on this:

> For us it is a security issue. Even if the whole source code of the app is public we keep some keys, like the signing key private. 

While it is theoretically possible to share signing keys with f-droid this was afaik never done for any app. Instead f-droid generates it's own signing keys for every app (~1800 currently I think) it distributes. In the android ecosystem that means that upgrading from one version to the other (upstream <--> fdroid) is not possible.

This is not ideal but normally not a problem you can just upgrade along the version you have chosen in the beginning.

>  So as far as I understand now the concept basically is to setup our own repo based on F-Droid server and publish our own app there, and that new repo you can add to your F-Droid client.

For how that could look for users see i.e. here: https://briarproject.org/fdroid.html

There is also a new project called [repomaker](https://f-droid.org/en/repomaker/) which simplifies the repository creation.

> an elegant solution to the security issue you mentioned would be to make fetlife build reproducible. The idea is that you create a recipe that can be used to create an identical apk from the source code.

That is in theory a very good solution to the signature and upgrade problem. Turns out though doing reproducible builds in android is quite hard still. So this realistically only works for the moment when both parties build the apk with the fdroidserver [build VM](https://f-droid.org/docs/Build_Server_Setup/).

But for now all talk about inclusion into the main fdroid repo is rather academical because this app contains quite a few non-free dependencies:

* crashlytics/fabric
* google firebase
* play-services-*
* google maps

I guess all of this could be made either optional or replaced by free alternatives but that's quite a bit of work I guess.

**Comment by Bubu** [comment]
@kistitan what's the situation between `FetLife` and `FetLifeKotlin`? These are two seperate implementations, right with the non-kotlin one being the current one? Does it make sense looking into the non-kotlin version or will it be replaced soon by the other one anyway?

**Comment by beatnickgr** [comment]
I'm happy to see there's no google services in the new app!
If I see it in F-Droid, I **promise** I will send some bitcoin (like I've sent already with other F-Droid apps).

**Comment by kistitan** [comment]
Still not any closer yet, due to the heavily used, not F-droid compatible features, and other priorities. 
We are on the other hand hoping to come up with a limited App version that can go into the Google Play Store

**Comment by FernandoMiguel** [comment]
> We are on the other hand hoping to come up with a limited App version that can go into the Google Play Store
how limited @kistitan ?

**Comment by kistitan** [comment]
Basically we are aiming for a FetChat app in the playstore, so it is planned to be a separate app, but hopefully we can aslo make it as a gateway to reach the full App, but we are still investigating both the technical options and also what the playstore policy allow us in that sense.

**Comment by Bubu** [comment]
Made in issue [here](https://gitlab.com/IzzyOnDroid/repo/issues/39), for inclusion in a third party f-droid repo.
